cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/OutboundHandler.java

```diff
@@ -39,9 +42,16 @@ public OutboundHandler(Channel inboundChannel) {
   }
 
   @Override
-  public void messageReceived(ChannelHandlerContext ctx, final MessageEvent e) throws Exception {
+  public void messageReceived(final ChannelHandlerContext ctx, final MessageEvent e) throws Exception {
     ChannelBuffer msg = (ChannelBuffer) e.getMessage();
-    inboundChannel.write(msg);
+
+    // TODO: can this cause messages to go out of order?
```

Comment:
Netty guarantees that the channel writes happen sequentially for a given channel. And the listeners are executed after a write, before the next channel write happens. So there will be no out of order messages.